business/authentication/openid_auth_controller.go

```diff
@@ -0,0 +1,1437 @@
+
+	// ParsedIdToken is the parsed form of the id_token, since it's known that it is a JWT.
+	ParsedIdToken *jwt.JSONWebToken
+
+	// IdTokenPayload holds the claims part of the id_token.
+	IdTokenPayload map[string]interface{}
+
+	// State is the code used to mitigate CSRF attacks.
+	State string
+
+	// Subject is the resolved username of the person that authenticated through an OpenId server.
+	Subject string
+
+	// UseAccessToken stores whether to use the OpenId access_token against the cluster API instead
+	// of the id_token.
+	UseAccessToken bool
+
+	// Error is nil unless there was an error during some phase of the authentication. A non-nil
+	// value cancels the authentication request.
+	Error error
+
+	// ShouldTerminateSession is set to a true value if an existing user session should be terminated
+	// as a consequence of a failure of a new authentication attempt (i.e if the Error field is not nil).
+	ShouldTerminateSession bool
+
+	// businessInstantiator is a function that returns an already initialized
+	// business layer. Normally, it should be set to the business.Get function.
+	// For tests, it can be set to something else that returns a compatible API.
+	businessInstantiator func(authInfo *api.AuthInfo) (*business.Layer, error)
+}
+
+// callbackCleanup deletes the nonce cookie that was generated during the redirection from Kiali to
+// the OpenId server to initiate authentication (see OpenIdAuthController.redirectToAuthServerHandler).
+func (p *openidFlowHelper) callbackCleanup(w http.ResponseWriter) *openidFlowHelper {
+	// Do nothing if there was an error in previous flow steps.
+	if p.Error != nil {
+		return p
+	}
+
+	// Delete the nonce cookie since we no longer need it.
+	deleteNonceCookie := http.Cookie{
+		Name:     OpenIdNonceCookieName,
+		Expires:  time.Unix(0, 0),
+		HttpOnly: true,
+		Path:     config.Get().Server.WebRoot,
+		SameSite: http.SameSiteStrictMode,
+		Value:    "",
+	}
+	http.SetCookie(w, &deleteNonceCookie)
+
+	return p
+}
+
+// extractOpenIdCallbackParams reads callback parameters from the HTTP request, once the OpenId server
+// redirects back to Kiali with the credentials. It also reads the nonce cookie with the code generated
+// during the initial redirection from Kiali to the OpenId Server (see OpenIdAuthController.redirectToAuthServerHandler).
+func (p *openidFlowHelper) extractOpenIdCallbackParams(r *http.Request) *openidFlowHelper {
+	// Do nothing if there was an error in previous flow steps.
+	if p.Error != nil {
+		return p
+	}
+
+	var err error
+
+	// Get the nonce code hash
+	var nonceCookie *http.Cookie
+	log.Errorf("Try to check nonce cookie")
+	if nonceCookie, err = r.Cookie(OpenIdNonceCookieName); err == nil {
+		log.Errorf("Nonce cookie = %v", nonceCookie.Value)
+		p.Nonce = nonceCookie.Value
+
+		hash := sha256.Sum224([]byte(nonceCookie.Value))
+		p.NonceHash = make([]byte, sha256.Size224)
+		copy(p.NonceHash, hash[:])
+	}
+	log.Errorf("Finish try to check nonce cookie")
```

Comment:
Are these debug msgs? Can they be removed?

Reply:
Oops! They are. I just removed them.